src/main/java/com/stablekernel/rxespresso/RxIdlingResource.java

```diff
@@ -129,12 +133,37 @@ public void onNext(T t) {
 
     private <T> void onFinally(Observable.OnSubscribe<T> onSubscribe, final String finalizeCaller) {
         int activeSubscriptionCount = subscriptions.decrementAndGet();
+        removeOnSubscribe(onSubscribe);
+
         if (LOG_LEVEL >= DEBUG) {
-            Log.d(TAG, onSubscribe + " - " + finalizeCaller + ": " + activeSubscriptionCount);
+            Log.d(TAG, "Subscription finished because: " + finalizeCaller + ". Active subscription count is now: " + activeSubscriptionCount);
+
+            if (LOG_LEVEL >= VERBOSE) {
+                // remaining subscriptions:
+                Log.d(TAG, ">>>>>>> remaining subscriptions ");
+                printStackTraces();
+                Log.d(TAG, "<<<<<<<<");
+            }
         }
         if (activeSubscriptionCount == 0) {
             Log.d(TAG, "onTransitionToIdle");
             resourceCallback.onTransitionToIdle();
         }
     }
+
+    private void removeOnSubscribe(Observable.OnSubscribe reference) {
+        for (WeakReference<Observable.OnSubscribe> weakRef : onSubscribeList.keySet()) {
```

Comment:
If this was a regular reference, you could just use the value as a key and remove it without looping.